apollo-router/src/plugins/router_limits/mod.rs

```diff
@@ -0,0 +1,366 @@
+//! A plugin for enforcing product limitations in the router based on License claims
+//!
+//! Currently includes:
+//! * TPS Rate Limiting: a certain threshold, set via License claim, for how many operations over a certain interval can be serviced
+
+use std::num::NonZeroU64;
+use std::time::Duration;
+
+use http::StatusCode;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use tower::limit::RateLimitLayer;
+use tower::load_shed::error::Overloaded;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use crate::graphql;
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::PluginInit;
+use crate::plugin::PluginPrivate;
+use crate::services::router;
+use crate::services::RouterResponse;
+
+#[derive(PartialEq, Debug, Clone, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+/// The limits placed on a router in virtue of what's in a user's license
+pub(crate) struct RouterLimits {
+    /// Transactions per second allowed based on license tier
+    pub(crate) tps: TpsLimitConf,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+/// Configuration for transactions per second
+pub(crate) struct TpsLimitConf {
+    /// The number of operations allowed during a certain interval
+    pub(crate) capacity: NonZeroU64,
+    /// The interval as specied in the user's license; this is in milliseconds
+    #[serde(deserialize_with = "humantime_serde::deserialize")]
+    #[schemars(with = "String")]
+    pub(crate) interval: Duration,
+}
+
+#[derive(Debug, Default, Deserialize, JsonSchema, Serialize)]
+pub(crate) struct RouterLimitsConfig {}
+
+#[async_trait::async_trait]
+impl PluginPrivate for RouterLimits {
+    type Config = RouterLimitsConfig;
+
+    // This will return an error only in cases where the router_limits plugin has been registered
+    // but there are no claims in the license for TPS. We _must_ check that there are claims in the
+    // router factory when regsitering this plugin
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let limits = init
+            .license
+            .get_limits()
+            .ok_or("License limits found during plugin initialization but failed to get limits in constructor phase of router_limits")
+            ?;
+
+        let tps = limits
+            .tps
+            .ok_or("License limits defined but no TPS claim defined")?;
+
+        let capacity = NonZeroU64::new(tps.capacity as u64)
+            .ok_or("Failed to convert TPS capacity into a usable value")?;
```

Comment:
Failure to init the plugin need further looking at. I'll need to check out the branch and have a look.